pkg/services/sqlstore/migrations/accesscontrol/permission_migrator.go

```diff
@@ -0,0 +1,233 @@
+package accesscontrol
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"xorm.io/xorm"
+
+	"github.com/grafana/grafana/pkg/services/accesscontrol"
+	"github.com/grafana/grafana/pkg/services/sqlstore/migrator"
+	"github.com/grafana/grafana/pkg/util"
+)
+
+var (
+	batchSize = 500
+)
+
+type permissionMigrator struct {
+	sess    *xorm.Session
+	dialect migrator.Dialect
+	migrator.MigrationBase
+}
+
+func (m *permissionMigrator) SQL(dialect migrator.Dialect) string {
+	return "code migration"
+}
+
+func (m *permissionMigrator) findRole(orgID int64, name string) (accesscontrol.Role, error) {
+	// check if role exists
+	var role accesscontrol.Role
+	_, err := m.sess.Table("role").Where("org_id = ? AND name = ?", orgID, name).Get(&role)
+	return role, err
+}
+
+func (m *permissionMigrator) bulkCreateRoles(allRoles []*accesscontrol.Role) ([]*accesscontrol.Role, error) {
+	if len(allRoles) == 0 {
+		return nil, nil
+	}
+
+	allCreatedRoles := make([]*accesscontrol.Role, 0, len(allRoles))
+
+	createRoles := m.createRoles
+	if m.dialect.DriverName() == migrator.MySQL {
+		createRoles = m.createRolesMySQL
+	}
+
+	// bulk role creations
+	err := batch(len(allRoles), batchSize, func(start, end int) error {
+		roles := allRoles[start:end]
+		createdRoles, err := createRoles(roles, start, end)
+		if err != nil {
+			return err
+		}
+		allCreatedRoles = append(allCreatedRoles, createdRoles...)
+		return nil
+	})
+
+	return allCreatedRoles, err
+}
+
+func (m *permissionMigrator) bulkAssignRoles(rolesMap map[int64]map[string]*accesscontrol.Role, assignments map[int64]map[string]struct{}) error {
+	if len(assignments) == 0 {
+		return nil
+	}
+
+	ts := time.Now()
+	userRoleAssignments := make([]accesscontrol.UserRole, 0)
+	teamRoleAssignments := make([]accesscontrol.TeamRole, 0)
+	builtInRoleAssignments := make([]accesscontrol.BuiltinRole, 0)
+
+	for orgID, roleNames := range assignments {
+		for name := range roleNames {
+			role, ok := rolesMap[orgID][name]
+			if !ok {
+				return &ErrUnknownRole{name}
+			}
+
+			if strings.HasPrefix(name, "managed:users") {
+				userID, err := strconv.ParseInt(strings.Split(name, ":")[2], 10, 64)
+				if err != nil {
+					return err
+				}
+				userRoleAssignments = append(userRoleAssignments, accesscontrol.UserRole{
+					OrgID:   role.OrgID,
+					RoleID:  role.ID,
+					UserID:  userID,
+					Created: ts,
```

Comment:
question: would we not want to set `Updated` field to the timestamp and not the `Created` field?